Avoid polling token endpoint after login request cancelled

adombeck opened the pull request:
If the context of the IsAuthenticated call from the PAM module to authd is cancelled, e.g. because the process of the PAM module exits, we want to automatically cancel the IsAuthenticated call to the broker, to avoid doing any further work towards authenticating the user (like sending requests to OAuth2 endpoints).

adombeck committed:
Simplify propagation of context cancellation to D-Bus broker

D-Bus does not support cancelling ongoing method calls through context
cancellation, in contrast to gRPC. So if the context of the
IsAuthenticated call from the PAM module to authd is cancelled (e.g.
because the PAM module exited unexpectedly), we need propagate that to
the broker by calling the CancelIsAuthenticated method.

We used to do that by calling IsAuthenticated in a goroutine and
waiting for the context to be cancelled in the main goroutine, calling
CancelIsAuthenticated when the context is cancelled.

This commit simplifies that, by passing the context to BusObject.CallWithContext, and calling
when the returned error is context.Cancelled.

This is only done for the D-Bus broker, not the example broker. That's
not a problem, because the example broker already cancels the
IsAuthenticated call if the context is cancelled, without requiring a
separate CancelIsAuthenticated call.

This commit also returns auth.Cancelled if the context was cancelled,
which was actually never used with the actual D-Bus broker, but with the
example broker and the testutils.BrokerBusMock.

adombeck commented:
I manually tested the changes by running `sudo login ...` and then cancelling the login via `Ctrl+C` when prompted to scan the QR code. The `IsAuthenticated` call to the broker is correctly cancelled, no more requests are sent to the token endpoint.

However, this comment: https://github.com/ubuntu/authd/blob/5863e338c2377e5cc6b18f0cbeb43096973480c2/internal/brokers/dbusbroker.go#L102-L102
... indicates that the previous behavior was on purpose. I did some investigations to find out the reason.

The functionality to cancel `IsAuthenticated` calls from authd to the broker via a separate `CancelIsAuthenticated` method was introduced in https://github.com/ubuntu/authd/pull/4 which closed UDENG-1107. That issue states "it was not possible to send a context through a dbus call". However, the godbus package does provide the `CallWithContext` method which automatically cancels the D-Bus call when the context is cancelled. I suspect that this was overlooked back then, and that there is not actually a good reason to not cancel the `IsAuthenticated` when the context is cancelled.

The first implementation actually used `CallWithContext` with the context, but that was changed in 53f1f92532b33ce4091cd6d9de5074e774af20a6. The comment itself was added later in 4b312ca06e7a1f6645e8bc47e287e649d96df224.

@denisonbarbosa, do you remember any more details? Am I missing something?

adombeck committed:
pam: Make the Go process exit when the parent process exits

The Go process started by the PAM module kept running even after the
process of the PAM module exited, causing authd to continue with the
authentication procedure.

adombeck commented:
Ok, I figured out what I was missing. The `Ctrl+C` to the PAM module causes an `EndSession` request being sent to the broker, which calls `CancelIsAuthenticated`. So the `IsAuthenticated` call is not cancelled because the context was cancelled on the client side, but because the client called `CancelIsAuthenticated`. The `CallWithContext` provided by the Go `dbus` module does not actually cause the call to be cancelled on the server side, it only cancels the request on the client side. I'll add a code comment which explains all that, so that we don't waste time on it again.

3v1n0 commented:
Yeah... Sorry for not mentioning it in the card, but I also did sole experiments with that after I wrote the initial thing, login disconnecting doesn't bring anything new to us.

So I feel we'll have to do process monitoring or something instead.

adombeck committed:
Implement Unwrap method of ToDisplayError

This is needed to be able to check the underlying error with errors.Is
or errors.As.

adombeck committed:
Explain why we need to call CancelIsAuthenticated

adombeck committed:
Fix inverted behavior of cancelFirstCall option

The first call was cancelled iff cancelFirstCall is false.

adombeck commented:
With the new commits I pushed, cancellation also works when I run `sudo login ...` and let it time out.

adombeck committed:
Improve error message

3v1n0 commented on pam/main-exec.go:
```diff
@@ -25,6 +27,19 @@ func init() {
 	runtime.LockOSThread()
+
+	// Ask the kernel to send SIGTERM if the parent dies
+	if err := unix.Prctl(unix.PR_SET_PDEATHSIG, uintptr(syscall.SIGTERM), 0, 0, 0); err != nil {
+		log.Errorf(context.Background(), "failed to set PDEATHSIG: %v", err)
+		os.Exit(1)
```
Yeah, I was wondering to use the same too... But maybe we can handle it also client-side, on the dbus side... Let me check that too.

However this is fine.

The changes on broker instead I'd hold since I had some refactoring there for another branch, and that side is quite delicate sadly (as we may cancel too early).

adombeck commented on internal/brokers/dbusbroker.go:
```diff
@@ -98,9 +99,16 @@ func (b dbusBroker) SelectAuthenticationMode(ctx context.Context, sessionID, aut
+		// BusObject.CallWithContext only aborts the call on the client side when the context is cancelled.
+		// So if that happens, we need to call CancelIsAuthenticated to also cancel the ongoing call on the broker side,
+		// to avoid that the broker continuous its work to authenticate the user.
+		b.CancelIsAuthenticated(context.Background(), sessionID)
+		return auth.Cancelled, "", nil
```
> `CancelIsAuthenticated` is not a blocking call

Really? It also uses `BusObject.CallWithContext` from the `dbus` package, which according to [the documentation](https://pkg.go.dev/github.com/godbus/dbus/v5#Object.Call) waits for a reply.

denisonbarbosa replied:
Yeah, the dbus call returns, but if you look at the `CancelIsAuthenticated` func on the broker side, it's not waiting for the cancellation to be done, it just requests a cancel
